File: apps/parcel/src/VietRide.Parcel.Application/Services/ParcelApprovalRequestedEvent.cs
using VietRide.Parcel.Application.Features.Parcels;
using VietRide.Shared.Application.Outbox;

namespace VietRide.Parcel.Application.Services;

internal static class ParcelApprovalRequestedEvent
{
    public static Task EnqueueAsync(
        IIntegrationEventOutbox outbox,
        Guid approvalRequestId,
        string requestType,
        Guid operatorId,
        Guid targetDriverUserId,
        Guid tripId,
        Guid? parcelId,
        Guid? incidentId,
        Guid? stopId,
        string validityCondition,
        DateTimeOffset occurredAt,
        CancellationToken cancellationToken)
    {
        var eventId = Guid.NewGuid();
        return ParcelOutboxEvents.EnqueueAsync(
            outbox,
            eventId,
            ParcelOutboxEvents.ApprovalRequested,
            new
            {
                eventId,
                occurredAt,
                approvalRequestId,
                requestType,
                operatorId,
                targetDriverUserId,
                tripId,
                parcelId,
                incidentId,
                stopId,
                expiresAt = (DateTimeOffset?)null,
                validityCondition,
                actionType = "OPEN_PARCEL_APPROVAL",
                actionParams = new { requestId = approvalRequestId, requestType },
            },
            cancellationToken);
    }
}

File: libs/shared/contracts/src/events/parcel-approval-requested.event.ts
import { z } from 'zod';

export const PARCEL_APPROVAL_REQUESTED_ROUTING_KEY = 'parcel.approval.requested';

export const ParcelApprovalRequestTypeSchema = z.enum([
  'CUSTODY_EXCEPTION',
  'STOP_DEPARTURE',
]);

export const ParcelApprovalRequestedEventSchema = z
  .object({
    eventId: z.string().uuid(),
    occurredAt: z.string().datetime({ offset: true }),
    approvalRequestId: z.string().uuid(),
    requestType: ParcelApprovalRequestTypeSchema,
    operatorId: z.string().uuid(),
    targetDriverUserId: z.string().uuid(),
    tripId: z.string().uuid(),
    parcelId: z.string().uuid().nullable(),
    incidentId: z.string().uuid().nullable(),
    stopId: z.string().uuid().nullable(),
    expiresAt: z.null(),
    validityCondition: z.string().trim().min(1),
    actionType: z.literal('OPEN_PARCEL_APPROVAL'),
    actionParams: z
      .object({
        requestId: z.string().uuid(),
        requestType: ParcelApprovalRequestTypeSchema,
      })
      .strict(),
  })
  .strict()
  .superRefine((event, context) => {
    if (event.actionParams.requestId !== event.approvalRequestId) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['actionParams', 'requestId'],
        message: 'Action requestId must match approvalRequestId.',
      });
    }
    if (event.actionParams.requestType !== event.requestType) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['actionParams', 'requestType'],
        message: 'Action requestType must match requestType.',
      });
    }
  });

export type ParcelApprovalRequestedEvent = z.infer<
  typeof ParcelApprovalRequestedEventSchema
>;
